### PR TITLE
Added upload checks

### LINKED ISSUE #593: Error Handling
Is it be possible to have an error handling when the file we send to the server is too heavy? At the moment, we have a 500 error without any error message 

In advance, I thank you

### PR DESCRIPTION
**Pull request by core23**
I am targetting this branch, because…

Rebase of #601 
Fixes #593 #600 #360
## Changelog

``` markdown
### Added
- Added filesize check to upload
- Added empty filename check to upload

```
## Subject

Added some checks while uploading a file.

### CODE REVIEW

**Comment by OskarStark** [comment]
@core23 can you please check travis?

**Comment by core23** [comment]
Looks like a travis problem. Can't reproduce this on my local machine

File: Provider/FileProvider.php [hunk 1]
```diff
@@ -310,14 +311,21 @@ public function validate(ErrorElement $errorElement, MediaInterface $media)
             throw new \RuntimeException(sprintf('Invalid binary content type: %s', get_class($media->getBinaryContent())));
         }
 
+        if ($media->getBinaryContent() instanceof UploadedFile && $media->getBinaryContent()->getClientSize() == 0) {
```
Comment (OskarStark) [inline Provider/FileProvider.php]:
```diff
- if ($media->getBinaryContent() instanceof UploadedFile && $media->getBinaryContent()->getClientSize() == 0) {
+ if ($media->getBinaryContent() instanceof UploadedFile && 0 == $media->getBinaryContent()->getClientSize()) {

File: Provider/FileProvider.php [hunk 2]
```diff
@@ -310,14 +311,21 @@ public function validate(ErrorElement $errorElement, MediaInterface $media)
...
+
         if (!in_array(strtolower(pathinfo($fileName, PATHINFO_EXTENSION)), $this->allowedExtensions)) {
             $errorElement
                 ->with('binaryContent')
                 ->addViolation('Invalid extensions')
                 ->end();
         }
 
-        if (!in_array($media->getBinaryContent()->getMimeType(), $this->allowedMimeTypes)) {
+        if ($media->getBinaryContent()->getFilename() != '' && !in_array($media->getBinaryContent()->getMimeType(), $this->allowedMimeTypes)) {
```
Comment (OskarStark) [inline Provider/FileProvider.php]:
```diff
- if ($media->getBinaryContent()->getFilename() != '' && !in_array($media->getBinaryContent()->getMimeType(), $this->allowedMimeTypes)) {
+ if ('' !=$media->getBinaryContent()->getFilename() && !in_array($media->getBinaryContent()->getMimeType(), $this->allowedMimeTypes)) {
```

File: Provider/FileProvider.php [hunk 3]
```diff
@@ -379,6 +387,12 @@ protected function doTransform(MediaInterface $media)
         $this->fixBinaryContent($media);
         $this->fixFilename($media);
 
+        if ($media->getBinaryContent() instanceof UploadedFile && $media->getBinaryContent()->getClientSize() == 0) {
```
Comment (OskarStark) [inline Provider/FileProvider.php]: yoda please

File: Provider/FileProvider.php [hunk 4]
```diff
@@ -379,6 +387,12 @@ protected function doTransform(MediaInterface $media)
...
+            $media->setProviderReference(sha1($media->getName().rand(11111, 99999)));
```
Comment (greg0ire) [inline Provider/FileProvider.php]: This is oddly specific… if you want a unique id, use `uniqid`, maybe?

**Comment by core23** [comment]
Fixed it

**Comment by core23** [comment]
Please review @OskarStark @greg0ire

**Comment by OskarStark** [comment]
Thank you @core23 👍

**Comment by jordisala1991** [comment]
You have to enable validation on the image field: `Assert\Valid()`. Im on my phone atm and not sure if this is documented (I believe it is), but I know 3 ways to do it:

Add validation on your entity. Using annotations or another format (xml, yaml)

There is a function on abstract admin that can do that also (search for validate on the class, or docs)

You can specify constraints on the field in the FormMapper.

**Comment by lemoinem** [comment]
Thank you for your answer!

I've just tried adding `Assert\Valid` to my $file property...
I'm not using SonataAdmin, I'm using my own form, so no FormMapper or abstract admin.

I can put my Entity and my FormType in a gist if you'd like. But, as of now, it looks like this feature is simply not working.

I tried adding a `Assert\Expression` to another field, and as long as my file is small enough not to generate the exception, I see the error.

As far as I can tell, the exception is thrown before the code ever has a chance to call validate. And it makes sense since `doTransform` is called way before any attempt to validate.

On the other hand, if I comment `throw new UploadException('The uploaded file is not found');` and `$media->setContentType($media->getBinaryContent()->getMimeType());` in `doTransform`, then I see the "The file is too big, max size: 2M" message.

I think `doTransform` should be rewritten so it never throws an exception.
At least, the call to `$provider->transform` in `ProviderDataTransformer.php` should be enclosed in a try-catch, maybe logging but definitely not rethrowing any exception.
Would you like me to provide a PR for this?

**Comment by jordisala1991** [comment]
Yes please, lets solve this issue. It will require some testing using sonata forms tho. Try to make ot BC on 3.x branch

**Pull request merged**